airbyte-integrations/connectors/source-teradata/src/main/java/io/airbyte/integrations/source/teradata/TeradataSource.java

```diff
@@ -0,0 +1,143 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TeradataSource.class);
+
+    private static final int INTERMEDIATE_STATE_EMISSION_FREQUENCY = 10_000;
+
+    static final String DRIVER_CLASS = "com.teradata.jdbc.TeraDriver";
+
+    public static final String PARAM_MODE = "mode";
+    public static final String PARAM_SSL = "ssl";
+    public static final String PARAM_SSL_MODE = "ssl_mode";
+    public static final String PARAM_SSLMODE = "sslmode";
+    public static final String PARAM_SSLCA = "sslca";
+    public static final String REQUIRE = "require";
+
+    private static final String CA_CERTIFICATE = "ca.pem";
+
+    public TeradataSource() {
+        super(DRIVER_CLASS, AdaptiveStreamingQueryConfig::new, new TeradataSourceOperations());
+    }
+
+    public static void main(final String[] args) throws Exception {
+        final Source source = new TeradataSource();
+        LOGGER.info("starting source: {}", TeradataSource.class);
+        new IntegrationRunner(source).run(args);
+        LOGGER.info("completed source: {}", TeradataSource.class);
+    }
+
+    @Override
+    public JsonNode toDatabaseConfig(final JsonNode config) {
+        final String schema = config.get(JdbcUtils.DATABASE_KEY).asText();
+
+        final String host = config.has(JdbcUtils.PORT_KEY) ?
+            config.get(JdbcUtils.HOST_KEY).asText() + ":" + config.get(JdbcUtils.PORT_KEY).asInt() :
+            config.get(JdbcUtils.HOST_KEY).asText();
+
+        final String jdbcUrl = String.format("jdbc:teradata://%s/", host);
+
+        final ImmutableMap.Builder<Object, Object> configBuilder = ImmutableMap.builder()
+            .put(JdbcUtils.USERNAME_KEY, config.get(JdbcUtils.USERNAME_KEY).asText())
+            .put(JdbcUtils.JDBC_URL_KEY, jdbcUrl)
+            .put(JdbcUtils.SCHEMA_KEY, schema);
+
+        if (config.has(JdbcUtils.PASSWORD_KEY)) {
+            configBuilder.put(JdbcUtils.PASSWORD_KEY, config.get(JdbcUtils.PASSWORD_KEY).asText());
+        }
+
+        if (config.has(JdbcUtils.JDBC_URL_PARAMS_KEY)) {
+            configBuilder.put(JdbcUtils.JDBC_URL_PARAMS_KEY, config.get(JdbcUtils.JDBC_URL_PARAMS_KEY).asText());
+        }
+
+        return Jsons.jsonNode(configBuilder.build());
+    }
+
+    @Override
+    public Set<String> getExcludedInternalNameSpaces() {
+        // the connector requires to have a database explicitly defined
+        return Set.of("");
+    }
+
+    @Override
+    protected int getStateEmissionFrequency() {
+        return INTERMEDIATE_STATE_EMISSION_FREQUENCY;
+    }
+
+    @Override
+    public List<TableInfo<CommonField<JDBCType>>> discoverInternal(JdbcDatabase database) throws Exception {
+        return discoverInternal(database, database.getSourceConfig().has(JdbcUtils.DATABASE_KEY) ?
+            database.getSourceConfig().get(JdbcUtils.DATABASE_KEY).asText() : null);
+    }
+
+    @Override
+    protected Map<String, String> getDefaultConnectionProperties(JsonNode config) {
```

Comment:
Getting a compile error on `override` for this method. For database sources you need to override `AbstractDBSource.toDatabaseConfig` method. See https://github.com/airbytehq/airbyte/blob/449b5def4bfff782b07167a5fa6077423b7b651e/airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java#L129 for reference.

Reply:
I have already overriden the `toDatabaseConfig()` and that is where I provide most of the config for the database (i.e. host, username, password, database, etc). `getDefaultConnectionProperties()` on the other hand was overriden because I didn't want to concatenate the additional connection parameters in the url myself and instead delegate that task to the db driver when creating the `DataSource` in `DataSourceFactory.build()`. I'm not sure why this capability was taken away and it was replaced with a static method call.

Reply:
`getDefaultConnectionProperties` is used to configure a Destination and not Source. You can remove this method as it is unused in `TeradataSource` class and keep `toDatabaseConfig`

Reply:
Yeah, but I still would have preferred to send a `Map<String, String>` to the DataSource and let the driver take care of constructing the connection parameters instead of me concatenating them it directly in the URL. 
Anyway... I will make the necessary change and ping you when I'm done.